DOCSP-47690: aws lambda tutorial (#207)
mongodb/docs-php-library: source/aws-lambda.txt

Locate the ``Serverless config`` name in the list of extensions provided
by the :github:`bref/extra-php-extension <brefphp/extra-php-extensions>`
package. Add it to the ``layers`` of the function in the ``serverless.yaml``
file, as shown in the following code:

.. code-block:: yaml

   plugins:
       - ./vendor/bref/bref
       - ./vendor/bref/extra-php-extensions # Adds the extra Serverless plugin

   functions:
       api:
           handler: index.php
           runtime: php-83-fpm
           layers:
               - ${bref-extra:mongodb-php-81} # Adds the MongoDB layer

Customize the Sample Application
--------------------------------

This step explains how to create a web page that list planets from the
Atlas :atlas:`sample data </sample-data>`. Replace the contents of
``index.php`` with the following code:

.. literalinclude:: /examples/aws-lambda/index.php
   :language: php

Redeploy the application with the new ``index.php`` by running the
following command:

.. code-block:: bash

   serverless deploy

The application page displays an error message because the ``MONGODB_URI``
environment variable is not set. The following section explains how to

> set your connection string in an environment variable.

Reviewer: This part got left in, just checking you didn't mean to delete it instead.

Author: Weird , yes meant to delete

.. _php-aws-lambda:

====================
Deploy to AWS Lambda
====================

.. facet::
   :name: genre
   :values: tutorial

.. meta::
   :keywords: serverless, deployment, code example, live

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the tool `Bref
<https://bref.sh>`__ to deploy serverless PHP applications on AWS
Lambda. This guide demonstrates how to deploy a PHP application built by
using the {+library-short+} and connect to an Atlas cluster by using AWS
IAM authentication.

Before You Get Started
----------------------

Before you can deploy to AWS Lambda by using Bref, you must set up the
following components:

- AWS account and access keys
- `Serverless Framework <https://www.serverless.com/>`__

The `Setup <https://bref.sh/docs/setup>`__ tutorial in the Bref
documentation describes how to prepare these components.

Install Dependencies
--------------------

Bref uses `Lambda layers
<https://docs.aws.amazon.com/lambda/latest/dg/chapter-layers.html>`__ to

> provide the PHP runtime. The ``bref`` layer is compiled with PHP and a

Author: Ill reword to make this more clear

       api:
           handler: index.php
           runtime: php-83-fpm
           layers:
               - ${bref-extra:mongodb-php-81} # Adds the MongoDB layer

Customize the Sample Application
--------------------------------

This step explains how to create a web page that list planets from the
Atlas :atlas:`sample data </sample-data>`. Replace the contents of
``index.php`` with the following code:

.. literalinclude:: /examples/aws-lambda/index.php
   :language: php

Redeploy the application with the new ``index.php`` by running the
following command:

.. code-block:: bash

   serverless deploy

The application page displays an error message because the ``MONGODB_URI``
environment variable is not set. The following section explains how to
set your connection string in an environment variable.

Set AWS Credentials
-------------------

Atlas supports passwordless authentication when using AWS credentials.
In any Lambda function, AWS sets environment variables that contain the
access token and secret token for the role assigned to deploy the function.

The following steps demonstrate how to set the role in your Atlas
cluster:

1. Open the Lambda function in the AWS console.

> #. Enter :guilabel:`Configuration > Permission`, then copy the :guilabel:`Role name`.

Reviewer: Q: When you say "Enter" is there a text entry field, or does Configuration > Permission refer to a page? If so, should this be "Select Configuration > Permission"?

Author: Navigate is probs more appropriate

#. Add this role to your Atlas cluster in the :guilabel:`Database
   Access` section. Select the :guilabel:`AWS IAM` authentication method
   and set the built-in role ``Read and write any database``.

To learn how to set up unified AWS access, see :atlas:`Set Up Unified
AWS Access </security/set-up-unified-aws-access/>` in the Atlas documentation.

After you configure the permissions, the Lambda function is allowed to access
your Atlas cluster. Next, configure your application to use the Atlas endpoint.

Access to Atlas clusters is also restricted by IP address. Since the
range of IP addresses that comes from AWS is very wide, you can allow
access from everywhere. To learn how to allow universal access, see
:atlas:`Configure IP Access List Entries </security/ip-access-list/>` in
the Atlas documentation.

.. note::

>    Using VPC Peering is recommended to isolate your Atlas
>    cluster from the internet. This requires the Lambda function to be
>    deployed in this AWS VPC.

Reviewer: Suggestion (also do you know what "this VPC" refers to?)

```suggestion
   We recommend using VPC Peering to isolate your Atlas
   cluster from the internet. You must deploy Lambda function to this AWS VPC
   to use VPC Peering.
```

Author: added some context

Next, copy your connection string and remove the ``<AWS access key>:<AWS
secret key>`` section, as your credentials are read from environment variables.

> Update the ``serverless.yml`` file in your project to set the
> environment variable ``MONGODB_URI`` to your connection string:

Reviewer: ```suggestion
Update the ``serverless.yml`` file in your project to set the ``MONGODB_URI`` environment variable to your connection string:
```

Author: reworded
